airflow/contrib/sensors/emr_run_job_flows.py

```diff
@@ -0,0 +1,185 @@
+    even when some fail.
+
+    :param job_flows: a queue of EMR JobFlows. It's a list of dicts, each one
+        mapping job_flow names to their configurations:
+        [{job_flow_name: job_flow_overrides}]. Each dict in the list represents
+        the job flows which should run in parallel, and every cluster in the
+        preceding dict is expected to have come to a successful terminal state,
+        prior to submitting the next dict. See boto3's job_flow_overrides EMR
+        details in
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/\
+        services/emr.html#EMR.Client.run_job_flow (templated)
+    :type job_flows: list
+    """
+
+    template_fields = ['job_flows']
+    template_ext = ()
+    # EMR logo... ~RGB(237,165,83)
+    ui_color = "#eda553"
+
+    # Overrides for EmrBaseSensor
+    NON_TERMINAL_STATES = EmrJobFlowSensor.NON_TERMINAL_STATES
+    FAILED_STATE = EmrJobFlowSensor.FAILED_STATE
+
+    @apply_defaults
+    def __init__(
+            self,
+            job_flows,
+            emr_conn_id='emr_default',
+            # require_auto_termination = False,
+            *args, **kwargs):
+        """
+        C0111
+        """
+        super().__init__(*args, **kwargs)
+        self.job_flows = job_flows
+        self.emr_conn_id = emr_conn_id
+        # These two fields will be filled in as clusters are requested and poked
+        self.current_batch = {}
+        self.statuses = []
+
+    def execute(self, context):
+        """
+        C0111
+        """
+        self.log.info(
+            "The clusters will be submitted across the following batches: %s",
+            [set(batch.keys()) for batch in self.job_flows])
+        # TODO: Verify all clusters set `"KeepJobFlowAliveWhenNoSteps": False`
+        # if self.require_auto_termination
+        return super().execute(context)
+
+    def get_emr_response(self):
+        """
+        override for EmrBaseSensor. Queries state of all clusters in current
+        batch and submits the next batch if they are all in a terminal state;
+        however, if any cluster is in a failed state, its state is returned so
+        that the current execution attempt fails.
+        """
+        emr_conn = EmrHook(emr_conn_id=self.emr_conn_id).get_conn()
+
+        responses = []
+        for name, job_flow_id in self.current_batch.items():
+            self.log.debug("Poking JobFlow {%s: %s}", name, job_flow_id)
+            response = emr_conn.describe_cluster(ClusterId=job_flow_id)
+            responses.append(response)
+            self._states()[name] = (job_flow_id, self._state_of(response))
+        self.log.debug("Poked JobFlow states: %s", self._states())
+
+        for failed in filter(lambda r: self._state_of(r) in
+                             EmrRunJobFlows.FAILED_STATE, responses):
+            self.log.info("there is at least one failed JobFlow")
+            return failed
```

Comment:
This is basically a find-first:

```python
failed_response = next(filter(lambda r: self._state_of(r) in EmrRunJobFlows.FAILED_STATE, responses), None)
if failed_response:
  self.log.info("there is at least one failed JobFlow")
  return failed_response
```

Reply:
Awesome! Glad it makes sense from reading it as is.
I had tried refactoring this and the below block of code (which looks for non_terminal responses) into a helper method, such as `_exists`, but it actually made the code harder to read.

Reply:
I mean your for loop is also a find first - it is not wrong. It is only because I think it is more logical (I would say) to use loops when looping over more than one item.

Reply:
I tried a few other approaches. I find this approach awkward as well, but it's used in some other languages, and short of moving to the [`:=` operator](https://www.python.org/dev/peps/pep-0572/#site-py) (not possible until airflow requires python >= v2.8)
I couldn't find a nicer way to do this - I am sure the `itertools` module has _something_ for this - but I don't think that's a much better solution and am unsure it'd even be supported in all versions of Python that Airflow expects.

Reply:
Oh. I'm sorry. I _just_ noticed your code snippet had an `if` instead of a `for`. Fair enough. I'll update to that. Thanks @feluelle

Reply:
Ah, my mistake. I _just_ noticed that your code snippet used an `if` instead of a `for`, and I was unaware that `next` had a `default` parameter. Thank you. I'll update to using that instead.